test/pettingzoo/pistonball_continuous.py

```diff
@@ -0,0 +1,234 @@
+                reward_normalization=args.rew_norm,
+                advantage_normalization=args.norm_adv,
+                recompute_advantage=args.recompute_adv,
+                # dual_clip=args.dual_clip,
+                # dual clip cause monotonically increasing log_std :)
+                value_clip=args.value_clip,
+                gae_lambda=args.gae_lambda,
+                action_space=env.action_space
+            )
+
+            agents.append(agent)
+            optims.append(optim)
+
+    policy = MultiAgentPolicyManager(
+        agents, env, action_scaling=True, action_bound_method='clip'
+    )
+    return policy, optims, env.agents
+
+
+def train_agent(
+    args: argparse.Namespace = get_args(),
+    agents: Optional[List[BasePolicy]] = None,
+    optims: Optional[List[torch.optim.Optimizer]] = None,
+) -> Tuple[dict, BasePolicy]:
+    train_envs = DummyVectorEnv([get_env for _ in range(args.training_num)])
+    test_envs = DummyVectorEnv([get_env for _ in range(args.test_num)])
+    # seed
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    train_envs.seed(args.seed)
+    test_envs.seed(args.seed)
+
+    policy, optim, agents = get_agents(args, agents=agents, optims=optims)
+
+    # collector
+    train_collector = Collector(
+        policy,
+        train_envs,
+        VectorReplayBuffer(args.buffer_size, len(train_envs)),
+        exploration_noise=False  # True
+    )
+    test_collector = Collector(policy, test_envs)
+    # train_collector.collect(n_step=args.batch_size * args.training_num)
+    # log
+    log_path = os.path.join(args.logdir, 'pistonball', 'dqn')
+    writer = SummaryWriter(log_path)
+    writer.add_text("args", str(args))
+    logger = TensorboardLogger(writer)
+
+    def save_fn(policy):
+        pass
+
+    def stop_fn(mean_rewards):
+        return False
+
+    def train_fn(epoch, env_step):
+        [agent.set_eps(args.eps_train) for agent in policy.policies.values()]
+
+    def test_fn(epoch, env_step):
+        [agent.set_eps(args.eps_test) for agent in policy.policies.values()]
+
+    def reward_metric(rews):
+        return rews[:, 0]
+
+    # trainer
+    result = onpolicy_trainer(
+        policy,
+        train_collector,
+        test_collector,
+        args.epoch,
+        args.step_per_epoch,
+        args.repeat_per_collect,
+        args.test_num,
+        args.batch_size,
+        episode_per_collect=args.episode_per_collect,
+        stop_fn=stop_fn,
+        save_fn=save_fn,
+        logger=logger,
+        resume_from_log=args.resume
+    )
+
+    return result, policy
+
+
+def watch(
+    args: argparse.Namespace = get_args(), policy: Optional[BasePolicy] = None
+) -> None:
+    env = get_env()
+    policy.eval()
+    collector = Collector(policy, env)
+    result = collector.collect(n_episode=1, render=args.render)
+    rews, lens = result["rews"], result["lens"]
+    print(f"Final reward: {rews[:, 0].mean()}, length: {lens.mean()}")

```

Comment:
```
Epoch #3: 4530it [00:07, 1167.98it/s, env_step=9591, len=283, n/ep=16, n/st=4530, piston_0/loss=138585331.676, piston_0/loss/clip=0.000, piston_0/loss/ent=1.415, piston_0/loss/vf=554341326.703, piston_1/loss=155627933.266, piston_1/loss/clip=0.000, piston_1/loss/ent=1.420, piston_1/Epoch #3: 4530it [00:07, 628.39it/s, env_step=9591, len=283, n/ep=16, n/st=4530, piston_0/loss=138585331.676, piston_0/loss/clip=0.000, piston_0/loss/ent=1.415, piston_0/loss/vf=554341326.703, piston_1/loss=155627933.266, piston_1/loss/clip=0.000, piston_1/loss/ent=1.420, piston_1/loss/vf=622511733.062, piston_2/loss=82814304.336, piston_2/loss/clip=-0.000, piston_2/loss/ent=1.421, piston_2/loss/vf=331257217.345, rew=-208.78] 
Epoch #3: test_reward: -131.375255 ± 557.657337, best_reward: -112.240233 ± 433.432150 in #0
```
It seems that the training is problematic... (huge loss)

Reply:
I rechecked this one. The main reason is the state space is a large image, and I used the `common/net` MLP model.
Should I define a convolution model in the test file?

Reply:
Sure, go ahead

Reply:
Sorry for the delay in giving an update. I'm currently behind on some personal schedules. 
I hope I can continue this from Jan 15.

Reply:
No problem at all, enjoy your vacation :)